src/option_learning.py

```diff
@@ -212,3 +220,206 @@ def update_segment_from_option_spec(self, segment: Segment,
                                   dtype=np.float32)
                 option = param_opt.ground([robby, block], params)
                 segment.set_option(option)
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class _LearnedSimpleParameterizedOption(ParameterizedOption):
+    """A convenience class for holding a learned parameterized option.
+
+    Prefer to use this because it is pickleable.
+    """
+    _operator: STRIPSOperator
+    _regressor: MLPRegressor
+    _changing_parameter_idxs: Sequence[int]
+
+    def __init__(self, name: str, operator: STRIPSOperator,
+                 regressor: MLPRegressor,
+                 changing_parameters: Sequence[Variable]) -> None:
+        assert set(changing_parameters).issubset(set(operator.parameters))
+        changing_parameter_idxs = [
+            i for i, v in enumerate(operator.parameters)
+            if v in changing_parameters
+        ]
+        types = [v.type for v in operator.parameters]
+        option_param_dim = sum(v.type.dim for v in changing_parameters)
+        params_space = Box(low=-np.inf,
+                           high=np.inf,
+                           shape=(option_param_dim, ),
+                           dtype=np.float32)
+        # Dataclass is frozen, so we have to do these hacks.
+        object.__setattr__(self, "_operator", operator)
+        object.__setattr__(self, "_regressor", regressor)
+        object.__setattr__(self, "_changing_parameter_idxs",
+                           changing_parameter_idxs)
+        super().__init__(name,
+                         types,
+                         params_space,
+                         _policy=self._regressor_based_policy,
+                         _initiable=self._precondition_based_initiable,
+                         _terminal=self._effect_based_terminal)
+
+    @cached_property
+    def _predicates(self) -> Set[Predicate]:
+        """Helper for initiable and terminal."""
+        return {a.predicate for a in self._operator.preconditions | \
+                self._operator.add_effects | self._operator.delete_effects}
+
+    def _precondition_based_initiable(self, state: State, memory: Dict,
+                                      objects: Sequence[Object],
+                                      params: Array) -> bool:
+        # The memory here is used to store the absolute params, based on
+        # the relative params and the object states.
+        memory["params"] = params  # store for sanity checking in policy
+        changing_objects = [objects[i] for i in self._changing_parameter_idxs]
+        memory["absolute_params"] = state.vec(changing_objects) + params
```

Comment:
What's going on here with absolute_params and relative_params? Is this a general algorithm or specific to some properties of your cover env?

Reply:
it's a general algorithm; hopefully the forthcoming changes to the docstring will make this clear